### PR TITLE
bank: active users in the system.

### LINKED ISSUE #111: bank: collect statistics for the bank activity
Users should be able to see statistics about the activity of the bank and the operations which are performed. Here are some metrics that need to be collected:

 * active users in the system

### PR DESCRIPTION
**Pull request by martinmilev**
Saving sessions in DB. Searching for sessions in the DB and counting the saved sessions.
Fixes #111.

### CODE REVIEW

**Comment by freekman** [comment]
is this PR ready for review

**Comment by freekman** [comment]
Note: The description of the PR would be better explained like
```js
The presence of a session is clear sign that a user is acto
```

**Comment by martinmilev** [comment]
PR is ready for review.

File: src/main/java/com/clouway/core/IdGenerator.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.clouway.core;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface IdGenerator {
```
Comment (freekman) [inline src/main/java/com/clouway/core/IdGenerator.java]: i think that instead of using this IdGenerator you could use the Provider

Comment (martinmilev) [inline src/main/java/com/clouway/core/IdGenerator.java]: ok.
done

File: src/main/java/com/clouway/core/Session.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package com.clouway.core;
+
+import java.util.Date;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class Session {
+  public final String userId;
```
Comment (freekman) [inline src/main/java/com/clouway/core/Session.java]: it is misleading id, userId .....

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  public PersistentSessionRepository(DataStore dataStore, IdGenerator idGenerator, MyClock clock) {
+    this.dataStore = dataStore;
+    this.idGenerator = idGenerator;
+    this.clock = clock;
+  }
+
+  @Override
+  public Session issueNewSession(String userId, Integer expireAfter) {
+    Date expiration = clock.getExpirationTime(expireAfter);
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java]: i am not sure that this is correct . I think that the ```SessionsRepository``` sould save(Session), get session and so on , what i am trying to say is that the creation of the sid and the calculation of the exp.time should be done elsewhere

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    String uuid = idGenerator.newId();
+
+    String query = "insert into sessions values(?,?,?)";
+    dataStore.update(query, uuid, userId, expiration);
+
+    return new Session(uuid, userId, expiration);
+  }
+
+  @Override
+  public Optional<Session> findSessionByUserId(String id) {
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java]: this would be better like findBySID()

File: src/main/java/com/clouway/core/SessionsRepository.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+   * @return Session
+   */
+  Optional<Session> findSessionByUserId(String userId);
+
+  /**
+   * Counts the current active sessions saved in the DB.
+   *
+   * @return the count of the sessions in Integer
+   */
+  Integer countActiveSessions();
```
Comment (freekman) [inline src/main/java/com/clouway/core/SessionsRepository.java]: this is misleading do you have unactive sessions

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    Session actual = repo.issueNewSession("John", 1);
+    Session expected = new Session("id", "John", date);
+
+    assertThat(actual, is(expected));
+  }
+
+  @Test
+  public void findSession() throws Exception {
+    DateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm");
```
Comment (freekman) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: are these needed why not assert on the session returned on register

File: src/main/resources/com/clouway/http/servlets/home.html [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+<body>
+<h1 class="text-center">myBank</h1>
+<hr>
+
+<div class="container-fluid col-md-2 col-md-offset-5">
+    <p class="bg-info text-center">
+        Active users: ${count}
+    </p>
+    <p class="bg-primary text-center">
+        Username: ${name}<br>
```
Comment (freekman) [inline src/main/resources/com/clouway/http/servlets/home.html]: you are continuing to use <br>

File: src/main/java/com/clouway/core/MyServerClock.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.clouway.core;
+
+import org.apache.commons.lang3.time.DateUtils;
+
+import java.util.Date;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class MyServerClock implements MyClock {
```
Comment (martinmilev) [inline src/main/java/com/clouway/core/MyServerClock.java]: done

File: src/main/java/com/clouway/core/Session.java (hunk 2 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/Session.java]: fixed.

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java (hunk 3 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java]: fixed.

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java (hunk 4 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepository.java]: fixed.

File: src/main/java/com/clouway/core/SessionsRepository.java (hunk 5 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsRepository.java]: fixed

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java (hunk 6 above)
Comment (martinmilev) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: Yes. MySQL returns incorrect time. It rounds some milliseconds.

**Comment by freekman** [comment]
Every time the same thing missing UI .How is this possible

File: src/main/java/com/clouway/http/servlets/HomePageServlet.java [hunk 9]
```diff
@@ -1,28 +1,62 @@
...
   public HomePageServlet() {
+    this(
+            new PersistentAccountRepository(new DataStore(new ConnectionProvider())),
+            new PersistentSessionRepository(new DataStore(new ConnectionProvider())),
+            new HtmlServletPageRenderer()
+    );
+  }
+
+  @VisibleForTesting
+  public HomePageServlet(AccountRepository repository, SessionsRepository sessions, ServletPageRenderer servletResponseWriter) {
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/HomePageServlet.java]: how is this a home page

File: src/main/java/com/clouway/http/servlets/LoginPageServlet.java [hunk 10]
```diff
@@ -17,30 +15,44 @@
...
+            new HtmlServletPageRenderer(),
+            new MyServerClock(),
+            new UuidGenerator(),
+            5
     );
   }
 
   @VisibleForTesting
-  public LoginPageServlet(AccountRepository repository, ServletPageRenderer servletResponseWriter) {
+  public LoginPageServlet(AccountRepository repository, SessionsRepository sessions, ServletPageRenderer servletResponseWriter, MyClock clock, Provider uuidGenerator, Integer sessionLife) {
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/LoginPageServlet.java]: here we have fare too many dependencies , more than 3 is a smell

File: src/main/java/com/clouway/core/SessionsRepository.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * Finds session in the DB by userID.
+   *
+   * @param uid id for the user who owns the new session
+   * @return Session
+   */
+  Optional<Session> findBySID(String uid);
+
+  Optional<Session> findByUserName(String uid);
```
Comment (freekman) [inline src/main/java/com/clouway/core/SessionsRepository.java]: why is the param named uid

File: src/main/java/com/clouway/core/SessionsRepository.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+  void deleteByUserName(String userName);
```
Comment (freekman) [inline src/main/java/com/clouway/core/SessionsRepository.java]: just deleteByName

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    repo.save(session);
+    Session actual = repo.findBySID("id").get();
+    Session expected = new Session("id", "John", date);
+
+    assertThat(actual, is(expected));
+  }
+
+  @Test
+  public void findSessionByUsername() throws Exception {
```
Comment (freekman) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: use camel case for naming

File: src/main/java/com/clouway/http/servlets/LoginPageServlet.java [hunk 14]
```diff
@@ -61,15 +74,25 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
...
     RegexValidator pswdValidator = new RegexValidator("[a-zA-Z_0-9]{6,18}");
 
-
     if (!pswd.equals(account.password) || !pswdValidator.check(pswd)) {
       servletResponseWriter.renderPage("login.html", Collections.singletonMap("error", "Wrong password"), resp);
     } else {
-      resp.addCookie(new Cookie("SID", "123"));
+      Optional<Session> possibleSession = sessions.findByUserName(name);
+
+      if (possibleSession.isPresent()) {
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/LoginPageServlet.java]: nooooo why is this here i dont like it the stupid filter has to deal firth the removing of the sessions but if @gadjevb and  @mgenov approve leave it i dont mind ,also i think you don't need findByUserName

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java (hunk 13 above)
Comment (martinmilev) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: fixed

File: src/main/java/com/clouway/core/SessionsRepository.java (hunk 12 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsRepository.java]: fixed

File: src/main/java/com/clouway/core/SessionsRepository.java (hunk 11 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsRepository.java]: fixed

**Comment by martinmilev** [comment]
Ready for review.

**Comment by freekman** [comment]
LGTM
